src/irr/video/IOpenGLPipeline.h

```diff
@@ -0,0 +1,91 @@
+        memset(m_uniformsSetForTheVeryFirstTime->data(), 0, m_uniformsSetForTheVeryFirstTime->size()*sizeof(bool));
+        memcpy(m_GLnames->data()+_ctxID*_STAGE_COUNT, _GLnames, _STAGE_COUNT*sizeof(GLuint));
+        std::move(_binaries, _binaries+_STAGE_COUNT, m_shaderBinaries->begin());
+
+        const size_t uVals_sz = _STAGE_COUNT*_ctxCount*IGPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE;
+        m_uniformValues = reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(uVals_sz, 128));
+    }
+    ~IOpenGLPipeline()
+    {
+        //shader programs can be shared so all names can be freed by any thread
+        for (GLuint n : *m_GLnames)
+            if (n != 0u)
+                COpenGLExtensionHandler::extGlDeleteProgram(n);
+        _IRR_ALIGNED_FREE(m_uniformValues);
+    }
+
+    bool haveUniformsBeenEverSet(uint32_t _stageIx, uint32_t _ctxID) const
+    {
+        const uint32_t ix = _ctxID*_STAGE_COUNT + _stageIx;
+        return !(*m_uniformsSetForTheVeryFirstTime)[ix];
+    }
+    void afterUniformsSet(uint32_t _stageIx, uint32_t _ctxID) const
+    {
+        const uint32_t ix = _ctxID*_STAGE_COUNT + _stageIx;
+        (*m_uniformsSetForTheVeryFirstTime)[ix] = false;
+    }
+
+    uint8_t* getPushConstantsStateForStage(uint32_t _stageIx, uint32_t _ctxID) const { return const_cast<uint8_t*>(m_uniformValues + ((_STAGE_COUNT*_ctxID + _stageIx)*IGPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE)); }
+
+    static core::smart_refctd_dynamic_array<GLint> gatherUniformLocations(const COpenGLSpecializedShader* _shader, GLuint _GLname)
+    {
+        auto uniforms_rng = _shader->getUniforms();
+
+        auto uniformLocations = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<GLint>>(uniforms_rng.length());
+        for (size_t i = 0ull; i < uniforms_rng.length(); ++i)
+            (*uniformLocations)[i] = COpenGLExtensionHandler::extGlGetUniformLocation(_GLname, uniforms_rng.begin()[i].m.name.c_str());
+
+        return uniformLocations;
+    }
+
+protected:
+    GLuint getShaderGLnameForCtx(uint32_t _stageIx, uint32_t _ctxID) const
+    {
+        const uint32_t name_ix = _ctxID*_STAGE_COUNT + _stageIx;
+
+        if ((*m_GLnames)[name_ix])
+            return (*m_GLnames)[name_ix];
+        else if ((*m_shaderBinaries)[_stageIx].binary)
+        {
+            const auto& bin = (*m_shaderBinaries)[_stageIx];
+
+            const GLuint GLname = COpenGLExtensionHandler::extGlCreateProgram();
+            COpenGLExtensionHandler::extGlProgramBinary(GLname, bin.format, bin.binary->data(), bin.binary->size());
+            (*m_GLnames)[_ctxID] = GLname;
+            return GLname;
+        }
+
+        return (*m_GLnames)[name_ix];
+    }
+
+    //mutable for deferred GL objects creation
+    mutable core::smart_refctd_dynamic_array<GLuint> m_GLnames;
+    mutable core::smart_refctd_dynamic_array<COpenGLSpecializedShader::SProgramBinary> m_shaderBinaries;
+    mutable core::smart_refctd_dynamic_array<bool> m_uniformsSetForTheVeryFirstTime;
```

Comment:
make these `std::array`, there's no need for dynamic allocation for arrays we 100% know the upper bound for.

Reply:
so what's the upper bound?

Reply:
argh shit `m_GLnames` and `m_uniformsSetForTheVeryFirstTime` actually needs a dynamic allocation (but you could join the `GLuint` and `bool` into a single struct/pair and only have one dynamically allocated array).

But `m_shaderBinaries` you only need up to MAX_STAGES_PER_PIPELINE (which is your template paramater right now).